### PR TITLE
Find CLI

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

This is a very simple charmhub find CLI, that displays the same
information that snap find does. Including how it handles missing search
queries.

I've yet to work out how to green tick the publisher, but that can come
later.

## QA steps

You need the following feature flag to test the new CLI.

`JUJU_DEV_FEATURE_FLAGS=charm-hub`

```sh
juju bootstrap lxd test
juju find wordpress
```

### CODE REVIEW

File: cmd/juju/charmhub/package_test.go [hunk 1]
```diff
@@ -1,16 +1,28 @@
...
+	"bytes"
 	"testing"
 
+	"github.com/juju/cmd"
+	jc "github.com/juju/testing/checkers"
 	gc "gopkg.in/check.v1"
 )
 
-//go:generate go run github.com/golang/mock/mockgen -package mocks -destination ./mocks/info_mock.go github.com/juju/juju/cmd/juju/charmhub InfoCommandAPI
+//go:generate go run github.com/golang/mock/mockgen -package mocks -destination ./mocks/info_mock.go github.com/juju/juju/cmd/juju/charmhub InfoCommandAPI,FindCommandAPI
```
Comment (hmlanigan) [inline cmd/juju/charmhub/package_test.go:15]: Suggest renaming the mocks/ file to better reflect contents

File: cmd/juju/charmhub/interface.go [hunk 2]
```diff
@@ -5,9 +5,24 @@ package charmhub
...
+
 // InfoCommandAPI describes API methods required
 // to execute the info command.
 type InfoCommandAPI interface {
 	Info(string) (charmhub.InfoResponse, error)
 	Close() error
 }
+
+// FindCommandAPI describes API methods required
+// to execute the info command.
```
Comment (hmlanigan) [inline cmd/juju/charmhub/interface.go:24]: s/info/find/

File: cmd/juju/charmhub/interface.go [hunk 3]
```diff
@@ -5,9 +5,24 @@ package charmhub
 
 import "github.com/juju/juju/api/charmhub"
 
+// Printer defines what is needed to print info.
```
Comment (hmlanigan) [inline cmd/juju/charmhub/interface.go:8]: s/print info./print./

File: cmd/juju/charmhub/find.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+	"github.com/juju/juju/api/charmhub"
+	jujucmd "github.com/juju/juju/cmd"
+	"github.com/juju/juju/cmd/modelcmd"
+)
+
+const (
+	findSummary = "Queries the charmhub store for available charms or bundles."
+	findDoc     = `
+The find command queries the charmhub store for available charms of bundles.
```
Comment (hmlanigan) [inline cmd/juju/charmhub/find.go:21]: s/of/or/

File: cmd/juju/charmhub/find.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	return jujucmd.Info(info)
+}
+
+// SetFlags defines flags which can be used with the find command.
+// It implements part of the cmd.Command interface.
+func (c *findCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.ModelCommandBase.SetFlags(f)
+	// TODO (stickupkid): add the following:
+	// --narrow
+	// --private
```
Comment (hmlanigan) [inline cmd/juju/charmhub/find.go:64]: we're not implementing --private, only --narrow, per doc

File: cmd/juju/charmhub/find.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+	return modelcmd.Wrap(&findCommand{})
+}
+
+// findCommand supplies the "find" CLI command used to display find information.
+type findCommand struct {
+	modelcmd.ModelCommandBase
+
+	api FindCommandAPI
+
+	verbose bool
```
Comment (hmlanigan) [inline cmd/juju/charmhub/find.go:42]: Where is this var used?  I believe we get --verbose for free from ModelCommandBase.

File: cmd/juju/charmhub/find_test.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/juju/juju/cmd/juju/charmhub/mocks"
+)
+
+type findSuite struct {
+	api *mocks.MockFindCommandAPI
+}
+
+var _ = gc.Suite(&findSuite{})
+
+func (s *findSuite) TestInitNoArgs(c *gc.C) {
```
Comment (hmlanigan) [inline cmd/juju/charmhub/find_test.go:21]: s/TestInitNoArgs/TestFindNoArgs/

File: cmd/juju/charmhub/find_test.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	// You can query the find api with no arguments.
+	command := &findCommand{}
+	err := command.Init([]string{})
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *findSuite) TestInitSuccess(c *gc.C) {
```
Comment (hmlanigan) [inline cmd/juju/charmhub/find_test.go:28]: s/TestInitSuccess/TestFindSuccess/

File: cmd/juju/charmhub/findwriter_test.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+	"github.com/juju/juju/api/charmhub"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+)
+
+type printFindSuite struct{}
+
+var _ = gc.Suite(&printFindSuite{})
+
+func (s *printFindSuite) TestCharmPrintFind(c *gc.C) {
```
Comment (hmlanigan) [inline cmd/juju/charmhub/findwriter_test.go:18]: Need a TestBundlePrintFind.

File: cmd/juju/charmhub/find_test.go (hunk 7 above)
Comment (SimonRichardson) [reply cmd/juju/charmhub/find_test.go:21]: `TestInit` is the function name :)

File: cmd/juju/charmhub/find_test.go (hunk 8 above)
Comment (SimonRichardson) [reply cmd/juju/charmhub/find_test.go:28]: `TestInit` is the method name :)

File: cmd/juju/charmhub/findwriter_test.go (hunk 9 above)
Comment (SimonRichardson) [reply cmd/juju/charmhub/findwriter_test.go:18]: I've updated to show bundles.

File: cmd/juju/charmhub/find_test.go (hunk 7 above)
Comment (hmlanigan) [reply cmd/juju/charmhub/find_test.go:21]: doh, :-)

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**